### DISCUSSION TITLE
Custom Normals and the Line Width node.

### DISCUSSION THREAD

**Discussion by dibli-goost** [bnpr/Malt#320] [Feedback & Feature Requests]
Putting this in feature request rather than issues because I don't necessarily think it's a bug, rather a limitation. The line width node currently takes the custom normals of a mesh into account when generating lines on an object, this can cause unwanted artifacts, like lines generated in the middle of a character's face.

With Custom Normals:
![image](https://user-images.githubusercontent.com/31767887/172718067-e354507f-c3b1-4e21-b655-75378093648e.png)
![image](https://user-images.githubusercontent.com/31767887/172718504-842722d2-5d27-4f5e-8819-df4933932b67.png)

Without:
![image](https://user-images.githubusercontent.com/31767887/172718152-05453390-9efe-480a-b6ff-95f54754601e.png)
![image](https://user-images.githubusercontent.com/31767887/172718710-88b9a137-c7dd-40cb-8083-f147eef42a3e.png)

We need a way to both retrieve the original, unedited normals of a mesh (I don't think that there is currently, I haven't found a way to), and a way to pass them through into the line width node.

Blend file example:
[Malt Line Width Normals.zip](https://github.com/bnpr/Malt/files/8865124/Malt.Line.Width.Normals.zip)

You can switch between the original and custom normals by turning the data transfer modifier off and on respectively.

**Comment by pragma37** [comment]
Ideally, the *depth* line detection should use the "true" non-smooth normals so this wouldn't be an issue.
It used to work that way but the way it caused some issues with alpha-clipped transparency.
So I disabled it as a "temp" fix and never looked back.
I'll look into it asap since custom normals + lines is a super common NPR workflow.

For *normal* line detection I think using custom normals would be the way to go in any case?

The *normal* line detection works based on the Pre Pass normal output so you can already override that.

However, that will also override the shading normal.
You can also pass any normal to shading nodes, but at the moment there's no built-in node to retrieve the originally loaded (non-modified) mesh normal from the Main Pass if you overrode it in the Pre Pass.

I'll add a built-in node for that. 

I'm not sure about loading both custom and non-custom normals, though. If you really need that for other use-cases, you can store them in a vertex color (directly or with [this feature](https://github.com/bnpr/Malt/discussions/142#discussioncomment-2361048) using geometry nodes.

**Reply by pragma37** [reply]
The line detection now uses the true normal and I've fixed all the edge cases (Development branch only for now).

I've also added variants to the surface functions to retrieve the original mesh attributes, so for example now we have a *surface normal* node and a *surface original normal* node.

Note, however, that the original normals are the ones that were loaded into the renderer before overriding them in the Pre Pass normal output, not the "non-custom" normals.
